### PR TITLE
[CWS] add a delay to hash file

### PR DESCRIPTION
**Pull request by safchain**
### What does this PR do?

- add a hash action report
- the hash report is used to introduce a hash delay in order to hash the file when the content is there. The file is hashed when the process exit or after a certain period of time
- the hash report is used to patch the event with the hashes data

### CODE REVIEW

File: pkg/security/secl/rules/ruleset.go [hunk 1]
```diff
@@ -284,6 +284,13 @@ func GetRuleEventType(rule *eval.Rule) (eval.EventType, error) {
 	return eventType, nil
 }
 
+func (rs *RuleSet) isActionAvailable(eventType eval.EventType, action *Action) bool {
+	if action.Def.Name() == HashAction && eventType != model.FileOpenEventType.String() {
+		return false
+	}
+	return true
+}
+
```
Comment (YoannGh) [inline pkg/security/secl/rules/ruleset.go:293]: With this change we won't be able to have a hash action on exec rules, should we also have this event type here as well?

File: pkg/security/secl/rules/ruleset.go [hunk 2]
```diff
@@ -148,8 +148,8 @@ func (rs *RuleSet) AddRules(parsingContext *ast.ParsingContext, pRules []*Policy
 	return result
 }
 
-// PopulateFieldsWithRuleActionsData populates the fields with the data from the rule actions
-func (rs *RuleSet) PopulateFieldsWithRuleActionsData(policyRules []*PolicyRule, opts PolicyLoaderOpts) *multierror.Error {
+// populateFieldsWithRuleActionsData populates the fields with the data from the rule actions
+func (rs *RuleSet) populateFieldsWithRuleActionsData(policyRules []*PolicyRule, opts PolicyLoaderOpts) *multierror.Error {
```
Comment (YoannGh) [inline pkg/security/secl/rules/ruleset.go:152]: `populateFieldsWithRuleActionsData` was made public by https://github.com/DataDog/datadog-agent/pull/28708 to fix the backend validation of rules, so I think we should keep it public for now.

File: pkg/security/secl/rules/ruleset.go (hunk 1 above)
Comment (safchain) [reply pkg/security/secl/rules/ruleset.go:293]: Good point

File: pkg/security/secl/rules/ruleset.go (hunk 2 above)
Comment (safchain) [reply pkg/security/secl/rules/ruleset.go:152]: Will revert this change

File: pkg/security/secl/rules/ruleset.go (hunk 1 above)
Comment (safchain) [reply pkg/security/secl/rules/ruleset.go]: done

File: pkg/security/secl/rules/ruleset.go (hunk 2 above)
Comment (safchain) [reply pkg/security/secl/rules/ruleset.go]: done

File: pkg/security/probe/file_hasher.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/DataDog/datadog-agent/pkg/security/secl/model"
+	"github.com/DataDog/datadog-agent/pkg/security/secl/rules"
+	"github.com/DataDog/datadog-agent/pkg/security/utils"
+)
+
+const (
+	defaultHashActionFlushDelay = 5 * time.Second
+)
+
+// FileHasher defines a process killer structure
```
Comment (spikat) [inline pkg/security/probe/file_hasher.go:27]: nit: not a process killer struct

File: pkg/security/probe/file_hasher.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// HashAndReport hash and report
+func (p *FileHasher) HashAndReport(rule *rules.Rule, ev *model.Event) {
+	eventType := ev.GetEventType()
+
+	// only open events are supported
+	if eventType != model.FileOpenEventType {
+		return
+	}
```
Comment (spikat) [inline pkg/security/probe/file_hasher.go:99]: Should we accept exec events here too (as for ruleset.go changes) ?

File: pkg/security/probe/probe_ebpf.go [hunk 5]
```diff
@@ -149,6 +149,9 @@ type EBPFProbe struct {
 	// On demand
```
Comment (spikat) [inline pkg/security/probe/probe_ebpf.go:149]: I think we should also modify other probes, right? (probe_ebpfless.go and even maybe windows?)

File: pkg/security/tests/action_test.go [hunk 6]
```diff
@@ -563,3 +563,74 @@ func TestActionKillDisarm(t *testing.T) {
...
+
+	t.Run("open", func(t *testing.T) {
+		test.WaitSignal(t, func() error {
+			go func() {
+				timeoutCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+
+				if err := runSyscallTesterFunc(
+					timeoutCtx, t, syscallTester,
+					"slow-write", "2", testFile, "aaa",
```
Comment (spikat) [inline pkg/security/tests/action_test.go:606]: If I understand correctly, here you test the case were the process writing the file exits BEFORE the 5s defaultHashActionFlushDelay ?
I wonder if we should also test the case where it takes more than this delay to exit?

File: pkg/security/probe/actions_linux.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+	Type  string `json:"type"`
+	Path  string `json:"path"`
+	State string `json:"state"`
+
+	// internal
+	resolved  bool
+	rule      *rules.Rule
+	pid       uint32
+	seenAt    time.Time
```
[the remark this answers is not in the document]
Comment (safchain) [reply pkg/security/probe/actions_linux.go:36]: that is event timestamp so yes we have it in the serialized event

File: pkg/security/probe/probe_ebpf.go (hunk 5 above)
Comment (safchain) [reply pkg/security/probe/probe_ebpf.go:149]: Windows we can't, but yes maybe ebpfless

File: pkg/security/probe/actions_linux.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (spikat) [reply pkg/security/probe/actions_linux.go:36]: Oh ok, I though the event timestamp will be set when the action is resolved, did we already serialize the time it took to resolve the actions ?

[the remark this answers is not in the document]
Comment (safchain) [reply pkg/security/probe/actions_linux.go:44]: no but I added a `trigger` to know the source of the action result

**Review by spikat — APPROVED** [review]

**Review by YoannGh — APPROVED** [review]

**Review by KevinFairise2 — APPROVED** [review]

**Pull request merged**